debian/manpages/mpeg2dec_onroot.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH MPEG2DEC_ONROOT 1 "June 17, 2005"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
mpeg2dec_onroot 
.SH SYNOPSIS
.B mpeg2dec_onroot
[-o <mode>] [-s [<track>]] [-t <pid>] [-c] [-w <id>] [-f <fps>] <file>
.SH OPTIONS
.TP
.B \-s
use program stream demultiplexer, track 0-15 or 0xe0-0xef
.TP
.B \-t
use transport stream demultiplexer, pid 0x10-0x1ffe
.TP
.B \-c
use c implementation, disables all accelerations
.TP
.B \-w      
window id
.TP
.B \-f      
target frame rate
.TP
.B \-o      
video output mode, can be: xv x11 null nullslice nullrgb16 nullrgb32 pgm pgmpipe md5
.br
.SH AUTHOR
This manual page was written by Alexis Sukrieh <sukria@sukria.net> 
for the Debian GNU/Linux system (but may be used by others).


debian/changelog
electricsheep (2.6.8-5) unstable; urgency=low

  * Removed the debian revision from the version string so sheeps can be
    downloaded from the server.
    (closes: #391990)

 -- Alexis Sukrieh <sukria@debian.org>  Wed, 11 Oct 2006 07:54:36 +0200

electricsheep (2.6.8-4) unstable; urgency=high

  [ Jérémy Bobbio ]

  * debian/patches/04_Composite.dpatch
    + tells mpeg2dec_onroot to skip ARGB visuals
    when trying to find the best visual. Fix an issue when running inside a
    Composite-enabled X environment.
    (closes: #390902)

  [ Alexis Sukrieh ]

  * urgency set to high, for closing a RC bug in testing (BSP).

 -- Alexis Sukrieh <sukria@debian.org>  Sat,  7 Oct 2006 16:49:00 +0200

electricsheep (2.6.8-3) unstable; urgency=low

  * mpeg2dec/libmpeg2/idct_altivec.c
    Applied patch from p2@mind.be for fixing the FTBFS issue on ppc
    architectures.
    (closes: #386673)

 -- Alexis Sukrieh <sukria@debian.org>  Mon, 18 Sep 2006 08:50:14 +0200

electricsheep (2.6.8-2) unstable; urgency=low

  [ The "Hey babe, I'm a Debian Developer!" upload, part #1 ] 
  
  * Changed Maintainer field in debian/control to `sukria@debian.org'. 

 -- Alexis Sukrieh <sukria@debian.org>  Sat,  9 Sep 2006 10:27:36 +0200

electricsheep (2.6.8-1) unstable; urgency=low

  * New upstream release.
    (closes: 365652)

 -- Alexis Sukrieh <sukria@sukria.net>  Sat,  5 Aug 2006 10:43:00 +0200

electricsheep (2.6.6-1) unstable; urgency=low

  * New upstream release.
  * Fixed some packaging issues reported by Philipp Kern.
  * Updated autotools scripts:
    + Added AM_MAINTAINER_MODE to configure.in
    + Customized Makefile.am so we can drop debian/patches/01_Makefile.dpatch
  * debian/rules now uses DESTDIR instead of PREFIX, cleaner (thanks to
    Philipp Kern).
  * Policy revision: 3.7.2

 -- Alexis Sukrieh <sukria@sukria.net>  Tue,  2 May 2006 00:28:02 +0200

electricsheep (2.6.4-2) unstable; urgency=low

  * Added missing build-dependencies: libx11-dev, libxext-dev, x-dev, libxt-dev 
    (closes: #344653)

 -- Alexis Sukrieh <sukria@sukria.net>  Sun, 25 Dec 2005 10:26:31 +0100

electricsheep (2.6.4-1) unstable; urgency=low

  * New upstream release.
    + possible to preview electricsheep in xscreensaver while running other
    instances (closes: #273984).
    + possible to run electricsheep in standalone mode, with a specific
    datadir (closes: #341004).
  * Removed debian/electricsheep.postinst, not needed anymore with the new
    upstream code tree.

 -- Alexis Sukrieh <sukria@sukria.net>  Wed, 21 Dec 2005 15:49:57 +0100

electricsheep (2.6.3+cvs20051206-1) unstable; urgency=low

  * New upstream release (CVS snapshot from 2005/12/06)
    + Fixes several buffer overflows
  * New patch for canceling the use of xsetbg for avoiding bug #325689.
  * New configure.in for using AM_MAINTAINER_MODE.

 -- Alexis Sukrieh <sukria@sukria.net>  Tue,  6 Dec 2005 10:42:23 +0100

electricsheep (2.6.3-2) unstable; urgency=low

  * Removed the dependency to xlibs-dev, added libxv-dev instead which is the
    only one needed here.
  * Remove stubs maintainer scripts (they came from the old maintenance and I
    forgot to remove them, let's do a bit of cleaning).
    - debian/prerem
    - debian/postrm
  * Remove debian/REJECT (was containing the licence).
  * Remove the overkilling dependency xlibs-dev, the other build-dependencies
    are enough.
  * Changed the autotools scripts, using AM_MAINTAINER_MODE, added a
    bootstrap.
  * Removed duplicate debian/electricsheep.desktop file, is provided in
    debian/default-files.

 -- Alexis Sukrieh <sukria@sukria.net>  Tue, 29 Nov 2005 21:00:56 +0100

electricsheep (2.6.3-1) unstable; urgency=low

  * New upstream release.
    (closes: #318003)
  * New patch for Makefile.in.
  * Added a .desktop file for KDE (Thanks to Gonéri Le Bouder).
    (closes: #315420)

 -- Alexis Sukrieh <sukria@sukria.net>  Mon, 25 Jul 2005 14:12:57 +0200

electricsheep (2.6.2-1) unstable; urgency=low

  * New upstream release: 2.6.2.
    (closes: #309852)
  * Added myself as maintainer (Sander Smeenk offered it to me).
  * Added dpatch for handling changes applied on upstream sources.
  * Added all missing manpages.

 -- Alexis Sukrieh <sukria@sukria.net>  Fri, 17 Jun 2005 18:28:17 +0200

electricsheep (2.5-2) unstable; urgency=low

  + Fixed builddepends on libpng (Closes: #245816)

 -- Sander Smeenk <ssmeenk@debian.org>  Sun, 25 Apr 2004 17:46:57 +0200

electricsheep (2.5-1) unstable; urgency=low

  + New Upstream Release (Closes: #243760)
  + vroot.h was updated which Closes: #217492
  
 -- Sander Smeenk <ssmeenk@debian.org>  Sun, 18 Apr 2004 15:06:34 +0200

electricsheep (2.4-3) unstable; urgency=low
  
  + Fixed dephelper depends (Closes: #216014)
  + Fixed conflicts with libmpeg2-3-dev (Closes: #207342)

 -- Sander Smeenk <ssmeenk@debian.org>  Sun, 19 Oct 2003 15:59:02 +0200

electricsheep (2.4-2) unstable; urgency=low

  + Correct dependencies (Closes: #206346)

 -- Sander Smeenk <ssmeenk@debian.org>  Sun, 24 Aug 2003 12:31:00 +0200
   
electricsheep (2.4-1) unstable; urgency=low

  + New upstream version (Closes: #203965)

 -- Sander Smeenk <ssmeenk@debian.org>  Thu, 14 Aug 2003 10:18:40 +0200

electricsheep (2.3-2) unstable; urgency=low

  + Moved *tif out of /usr/share to /usr/share/electricsheep.
  + Forgot to close bugs with my previous upload.
    Closes: #192510
  + Buildproblems on noni386 are fixed upstream
    Thanks Julien Lemoine for your effort.
    Closes: #167963, #182510
  + Added build dependency to libexpat1-dev (Closes: #185390)
  + Actually using CFLAGS from debian/rules now

 -- Sander Smeenk <ssmeenk@debian.org>  Tue, 18 Mar 2003 14:44:08 +0100

electricsheep (2.3-1) unstable; urgency=low

  + New upstream.
  + New debian/ structure, new Standards version, etc.

 -- Sander Smeenk <ssmeenk@debian.org>  Wed, 26 Feb 2003 09:26:08 +0100

electricsheep (2.1-6) unstable; urgency=low

  + 'electricsheep.org' the domain that pointed to the official

mpeg2dec/src/mpeg2dec_onroot.c
/*
 * mpeg2dec.c
 * Copyright (C) 2000-2002 Michel Lespinasse <walken@zoy.org>
 * Copyright (C) 1999-2000 Aaron Holtzman <aholtzma@ess.engr.uvic.ca>
 *
 * This file is part of mpeg2dec, a free MPEG-2 video stream decoder.
 * See http://libmpeg2.sourceforge.net/ for updates.
 *
 * mpeg2dec is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * mpeg2dec is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */


#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <getopt.h>
#ifdef HAVE_SYS_TIME_H
#include <sys/time.h>
#include <signal.h>
#endif
#include <inttypes.h>

#include "video_out.h"
#include "mpeg2.h"
#include "mm_accel.h"

#define BUFFER_SIZE 4096
static uint8_t buffer[BUFFER_SIZE];
static FILE * in_file;
static int demux_track = 0;
static int demux_pid = 0;
static int disable_accel = 0;
static mpeg2dec_t mpeg2dec;
static vo_open_t * output_open = NULL;

int window_id = -2;
int quiet = 1;
int target_fps = 23;

#ifdef HAVE_SYS_TIME_H

static void print_fps (int final);

static RETSIGTYPE signal_handler (int sig)
{
    print_fps (1);
    signal (sig, SIG_DFL);
    raise (sig);
}

static void print_fps (int final) 
{
    static uint32_t frame_counter = 0;
    static struct timeval tv_beg, tv_start, tv_rate;
    static int total_elapsed;
    static int last_count = 0;
    struct timeval tv_end;
    int fps, tfps, frames, elapsed;

    gettimeofday (&tv_end, NULL);

    if (!frame_counter) {
	tv_rate = tv_start = tv_beg = tv_end;
	signal (SIGINT, signal_handler);
    }

    elapsed = (tv_end.tv_sec - tv_beg.tv_sec) * 100 +
	(tv_end.tv_usec - tv_beg.tv_usec) / 10000;
    total_elapsed = (tv_end.tv_sec - tv_start.tv_sec) * 100 +
	(tv_end.tv_usec - tv_start.tv_usec) / 10000;

    if (final) {
	if (total_elapsed) 
	    tfps = frame_counter * 10000 / total_elapsed;
	else
	    tfps = 0;

	if (!quiet)
	  fprintf (stderr,"\n%d frames decoded in %d.%02d "
		   "seconds (%d.%02d fps)\n", frame_counter,
		   total_elapsed / 100, total_elapsed % 100,
		   tfps / 100, tfps % 100);

	return;
    }

    frame_counter++;

    if (elapsed < 50)	/* only display every 0.50 seconds */
	return;

    tv_beg = tv_end;
    frames = frame_counter - last_count;

    fps = frames * 10000 / elapsed;			/* 100x */
    tfps = frame_counter * 10000 / total_elapsed;	/* 100x */

    if (!quiet)
      fprintf (stderr, "%d frames in %d.%02d sec (%d.%02d fps), "
	       "%d last %d.%02d sec (%d.%02d fps)\n", frame_counter,
	       total_elapsed / 100, total_elapsed % 100,
	       tfps / 100, tfps % 100, frames, elapsed / 100, elapsed % 100,
	       fps / 100, fps % 100);

    last_count = frame_counter;

}

#else /* !HAVE_SYS_TIME_H */

static void print_fps (int final)
{
}

#endif

static void print_usage (char ** argv)
{
    int i;
    vo_driver_t * drivers;

    fprintf (stderr,
	     "usage: %s [-o <mode>] [-s [<track>]] [-t <pid>] [-c] "
	     "[-w <id>] [-f <fps>] <file>\n"
	     "\t-s\tuse program stream demultiplexer, "
	     "track 0-15 or 0xe0-0xef\n"
	     "\t-t\tuse transport stream demultiplexer, pid 0x10-0x1ffe\n"
	     "\t-c\tuse c implementation, disables all accelerations\n"
	     "\t-w\twindow id\n"
	     "\t-f\ttarget frame rate\n"
	     "\t-o\tvideo output mode\n", argv[0]);

    drivers = vo_drivers ();
    for (i = 0; drivers[i].name; i++)
	fprintf (stderr, "\t\t\t%s\n", drivers[i].name);

    exit (1);
}

static void handle_args (int argc, char ** argv)
{
    int c;
    vo_driver_t * drivers;
    int i;
    char * s;

    drivers = vo_drivers ();
    while ((c = getopt (argc, argv, "s::t:co:w:f:")) != -1)
	switch (c) {
	case 'o':
	    for (i = 0; drivers[i].name != NULL; i++)
		if (strcmp (drivers[i].name, optarg) == 0)
		    output_open = drivers[i].open;
	    if (output_open == NULL) {
		fprintf (stderr, "Invalid video driver: %s\n", optarg);
		print_usage (argv);
	    }
	    break;

	case 's':
	    demux_track = 0xe0;
	    if (optarg != NULL) {
		demux_track = strtol (optarg, &s, 16);
		if (demux_track < 0xe0)
		    demux_track += 0xe0;
		if ((demux_track < 0xe0) || (demux_track > 0xef) || (*s)) {
		    fprintf (stderr, "Invalid track number: %s\n", optarg);
		    print_usage (argv);
		}
	    }
	    break;

	case 't':
	    demux_pid = strtol (optarg, &s, 16);
	    if ((demux_pid < 0x10) || (demux_pid > 0x1ffe) || (*s)) {
		fprintf (stderr, "Invalid pid: %s\n", optarg);
		print_usage (argv);
	    }
	    break;

	case 'c':
	    disable_accel = 1;
	    break;

	case 'w':
	  window_id = strtol (optarg, &s, 0);
	  break;

	case 'f':
	  target_fps = strtol (optarg, &s, 10);
	  if (0 >= target_fps) {
	    fprintf(stderr, "target frame rate must be positive, not %d.\n", target_fps);
	    print_usage(argv);
	  }
	  break;

	default:
	    print_usage (argv);
	}

    /* -o not specified, use a default driver */
    if (output_open == NULL)
	output_open = drivers[0].open;

    if (optind < argc) {
	in_file = fopen (argv[optind], "rb");
	if (!in_file) {
	    fprintf (stderr, "%s - couldnt open file %s\n", strerror (errno),
		     argv[optind]);
	    exit (1);
	}
    } else
	in_file = stdin;
}

static void decode_mpeg2 (uint8_t * buf, uint8_t * end)
{
    int num_frames;

    num_frames = mpeg2_decode_data (&mpeg2dec, buf, end);
    while (num_frames--)
	print_fps (0);
}

#define DEMUX_PAYLOAD_START 1
static int demux (uint8_t * buf, uint8_t * end, int flags)
{
    static int mpeg1_skip_table[16] = {
	0, 0, 4, 9, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
    };

    /*
     * the demuxer keeps some state between calls:
     * if "state" = DEMUX_HEADER, then "head_buf" contains the first
     *     "bytes" bytes from some header.
     * if "state" == DEMUX_DATA, then we need to copy "bytes" bytes
     *     of ES data before the next header.
     * if "state" == DEMUX_SKIP, then we need to skip "bytes" bytes
     *     of data before the next header.
     *
     * NEEDBYTES makes sure we have the requested number of bytes for a
     * header. If we dont, it copies what we have into head_buf and returns,
     * so that when we come back with more data we finish decoding this header.
     *
     * DONEBYTES updates "buf" to point after the header we just parsed.
     */

#define DEMUX_HEADER 0
#define DEMUX_DATA 1
#define DEMUX_SKIP 2
    static int state = DEMUX_SKIP;
    static int state_bytes = 0;
    static uint8_t head_buf[264];

    uint8_t * header;
    int bytes;
    int len;

#define NEEDBYTES(x)						\
    do {							\
	int missing;						\
								\
	missing = (x) - bytes;					\
	if (missing > 0) {					\
	    if (header == head_buf) {				\
		if (missing <= end - buf) {			\
		    memcpy (header + bytes, buf, missing);	\
		    buf += missing;				\
		    bytes = (x);				\
		} else {					\
		    memcpy (header + bytes, buf, end - buf);	\
		    state_bytes = bytes + end - buf;		\
		    return 0;					\
		}						\
	    } else {						\
		memcpy (head_buf, header, bytes);		\
		state = DEMUX_HEADER;				\
		state_bytes = bytes;				\
		return 0;					\
	    }							\
	}							\
    } while (0)

#define DONEBYTES(x)		\
    do {			\
	if (header != head_buf)	\
	    buf = header + (x);	\
    } while (0)

    if (flags & DEMUX_PAYLOAD_START)
	goto payload_start;
    switch (state) {
    case DEMUX_HEADER:
	if (state_bytes > 0) {
	    header = head_buf;
	    bytes = state_bytes;
	    goto continue_header;
	}
	break;
    case DEMUX_DATA:
	if (demux_pid || (state_bytes > end - buf)) {
	    decode_mpeg2 (buf, end);
	    state_bytes -= end - buf;
	    return 0;
	}
	decode_mpeg2 (buf, buf + state_bytes);
	buf += state_bytes;
	break;
    case DEMUX_SKIP:
	if (demux_pid || (state_bytes > end - buf)) {
	    state_bytes -= end - buf;
	    return 0;
	}
	buf += state_bytes;
	break;
    }

    while (1) {
	if (demux_pid) {
	    state = DEMUX_SKIP;
	    return 0;
	}
    payload_start:
	header = buf;
	bytes = end - buf;
    continue_header:
	NEEDBYTES (4);
	if (header[0] || header[1] || (header[2] != 1)) {
	    if (demux_pid) {
		state = DEMUX_SKIP;
		return 0;
	    } else if (header != head_buf) {
		buf++;
		goto payload_start;
	    } else {
		header[0] = header[1];
		header[1] = header[2];
		header[2] = header[3];
		bytes = 3;
		goto continue_header;
	    }
	}
	if (demux_pid) {
	    if ((header[3] >= 0xe0) && (header[3] <= 0xef))
		goto pes;
	    fprintf (stderr, "bad stream id %x\n", header[3]);
	    exit (1);
	}
	switch (header[3]) {
	case 0xb9:	/* program end code */
	    /* DONEBYTES (4); */
	    /* break;         */
	    return 1;
	case 0xba:	/* pack header */
	    NEEDBYTES (12);
	    if ((header[4] & 0xc0) == 0x40) {	/* mpeg2 */
		NEEDBYTES (14);
		len = 14 + (header[13] & 7);
		NEEDBYTES (len);
		DONEBYTES (len);
		/* header points to the mpeg2 pack header */
	    } else if ((header[4] & 0xf0) == 0x20) {	/* mpeg1 */
		DONEBYTES (12);
		/* header points to the mpeg1 pack header */
	    } else {
		fprintf (stderr, "weird pack header\n");
		exit (1);
	    }
	    break;
	default:
	    if (header[3] == demux_track) {
	    pes:
		NEEDBYTES (7);
		if ((header[6] & 0xc0) == 0x80) {	/* mpeg2 */
		    NEEDBYTES (9);
		    len = 9 + header[8];
		    NEEDBYTES (len);
		    /* header points to the mpeg2 pes header */
		    if (header[7] & 0x80) {
			uint32_t pts;

			pts = (((buf[9] >> 1) << 30) |
			       (buf[10] << 22) | ((buf[11] >> 1) << 15) |
			       (buf[12] << 7) | (buf[13] >> 1));
			mpeg2_pts (&mpeg2dec, pts);
		    }
		} else {	/* mpeg1 */
		    int len_skip;
		    uint8_t * ptsbuf;

		    len = 7;
		    while (header[len - 1] == 0xff) {
			len++;
			NEEDBYTES (len);
			if (len == 23) {
			    fprintf (stderr, "too much stuffing\n");
			    break;
			}
		    }
		    if ((header[len - 1] & 0xc0) == 0x40) {
			len += 2;
			NEEDBYTES (len);
		    }
		    len_skip = len;
		    len += mpeg1_skip_table[header[len - 1] >> 4];
		    NEEDBYTES (len);
		    /* header points to the mpeg1 pes header */
		    ptsbuf = header + len_skip;
		    if (ptsbuf[-1] & 0x20) {
			uint32_t pts;

			pts = (((ptsbuf[-1] >> 1) << 30) |
			       (ptsbuf[0] << 22) | ((ptsbuf[1] >> 1) << 15) |
			       (ptsbuf[2] << 7) | (ptsbuf[3] >> 1));
			mpeg2_pts (&mpeg2dec, pts);
		    }
		}
		DONEBYTES (len);
		bytes = 6 + (header[4] << 8) + header[5] - len;
		if (demux_pid || (bytes > end - buf)) {
		    decode_mpeg2 (buf, end);
		    state = DEMUX_DATA;
		    state_bytes = bytes - (end - buf);
		    return 0;
		} else if (bytes > 0) {
		    decode_mpeg2 (buf, buf + bytes);
		    buf += bytes;
		}
	    } else if (header[3] < 0xb9) {
		fprintf (stderr,
			 "looks like a video stream, not system stream\n");
		exit (1);
	    } else {
		NEEDBYTES (6);
		DONEBYTES (6);
		bytes = (header[4] << 8) + header[5];
		if (bytes > end - buf) {
		    state = DEMUX_SKIP;
		    state_bytes = bytes - (end - buf);
		    return 0;
		}
		buf += bytes;
	    }
	}
    }
}

static void ps_loop (void)
{
    uint8_t * end;

    do {
	end = buffer + fread (buffer, 1, BUFFER_SIZE, in_file);
	if (demux (buffer, end, 0))
	    break;	/* hit program_end_code */
    } while (end == buffer + BUFFER_SIZE);
}

static void ts_loop (void)
{
#define PACKETS (BUFFER_SIZE / 188)
    uint8_t * buf;
    uint8_t * data;
    uint8_t * end;
    int packets;
    int i;
    int pid;

    do {
	packets = fread (buffer, 188, PACKETS, in_file);
	for (i = 0; i < packets; i++) {
	    buf = buffer + i * 188;
	    end = buf + 188;
	    if (buf[0] != 0x47) {
		fprintf (stderr, "bad sync byte\n");
		exit (1);
	    }
	    pid = ((buf[1] << 8) + buf[2]) & 0x1fff;
	    if (pid != demux_pid)
		continue;
	    data = buf + 4;
	    if (buf[3] & 0x20) {	/* buf contains an adaptation field */
		data = buf + 5 + buf[4];
		if (data > end)
		    continue;
	    }
	    if (buf[3] & 0x10)
		demux (data, end, (buf[1] & 0x40) ? DEMUX_PAYLOAD_START : 0);
	}
    } while (packets == PACKETS);
}

static void es_loop (void)
{
    uint8_t * end;
		
    do {
	end = buffer + fread (buffer, 1, BUFFER_SIZE, in_file);
	decode_mpeg2 (buffer, end);
    } while (end == buffer + BUFFER_SIZE);
}

int main (int argc, char ** argv)
{
    vo_instance_t * output;
    uint32_t accel;

    if (!quiet) fprintf (stderr, PACKAGE"-"VERSION
	     " - by Michel Lespinasse <walken@zoy.org> and Aaron Holtzman\n");
    handle_args (argc, argv);

    accel = disable_accel ? 0 : (mm_accel () | MM_ACCEL_MLIB);

    vo_accel (accel);
    output = vo_open (output_open);
    if (output == NULL) {
	fprintf (stderr, "Can not open output\n");
	return 1;
    }
    mpeg2_init (&mpeg2dec, accel, output);

    if (demux_pid)
	ts_loop ();
    else if (demux_track)
	ps_loop ();
    else
	es_loop ();

    mpeg2_close (&mpeg2dec);
    vo_close (output);
    print_fps (1);
    return 0;
}
